enterprise/backend/src/metabase_enterprise/dependencies/findings.clj

```diff
@@ -0,0 +1,71 @@
+(ns metabase-enterprise.dependencies.findings
+  (:require
+   [metabase-enterprise.dependencies.analysis :as deps.analysis]
+   [metabase-enterprise.dependencies.models.analysis-finding :as deps.analysis-finding]
+   [metabase.lib-be.core :as lib-be]
+   [metabase.lib.core :as lib]
+   [metabase.util.log :as log]
+   [metabase.util.malli :as mu]
+   [toucan2.core :as t2]))
+
+(set! *warn-on-reflection* true)
+
+(def ^:private model->dependency-type
+  {:model/Card :card
+   :model/Transform :transform
+   :model/Segment :segment})
+
+(defmulti ^:private get-db-id
+  "Gets the database id for a toucan instance"
+  {:arglists '([toucan-instance])}
+  t2/model)
+
+(defmethod get-db-id :default
+  [toucan-instance]
+  (:database_id toucan-instance))
+
+(defmethod get-db-id :model/Transform
+  [toucan-instance]
+  (some-> toucan-instance :source :query lib/database-id))
+
+(defmethod get-db-id :model/Segment
+  [toucan-instance]
+  (some-> toucan-instance :definition lib/database-id))
+
+(defn upsert-analysis!
+  "Given a Toucan entity, run its analysis and write the results into `:model/AnalysisFinding`.
+
+  If any row exists already, it is replaced. If it does not exist, it is created."
+  [toucan-instance]
+  (when-not (lib-be/metadata-provider-cache)
+    (log/warn "FIXME: deps.findings/upsert-analysis! ran without reusing `MetadataProvider`s"))
```

Comment:
`log/warn` like this, if there is a case where it's called, is just going to spam the logs in Cloud and make the Cloud team mad at us... have this throw and then fix the places where it throws.

Btw the metadata provider cache is bound automatically for every request these days